mss/L2Cache.cpp

```diff
@@ -0,0 +1,657 @@
+
+        // Access cache, and check cache hit or miss
+        if (req->getCacheState() == L2CacheState::RELOAD) {
+            
+            if (cacheLookUpResult == L2CacheState::MISS) {
+                
+                // Reload cache line
+                reloadCache_(req->getInstPtr()->getRAdr());
+                
+                ILOG("Reload Complete: phyAddr=0x" << std::hex << req->getInstPtr()->getRAdr());
+            }
+            
+            req->setCacheState(L2CacheState::HIT);
+        }
+        else {
+        
+            // Update memory access info
+            req->setCacheState(cacheLookUpResult);
+        }
+    }
+
+    // Pipeline Stage HIT_MISS_HANDLING
+    void L2Cache::handleCacheAccessResult_() {
+        const auto req = l2cache_pipeline_[stages_.HIT_MISS_HANDLING];
+        ILOG("Pipeline stage HIT_MISS_HANDLING : " << req->getInstPtr());
+        
+        inFlight_reqs_--;
+        
+        // This request to access cache came from DCache or ICache to do a cache lookup.
+        // It was either a miss or hit based on cacheLookup_() in the previous stage of the pipeline
+        if (req->getCacheState() == L2CacheState::HIT) {
+            // If it was originally a miss in L2Cache, on return from BIU, it's SrcUnit is set to BIU
+            // and DestUnit to whatever the original SrcUnit was.
+            //
+            // If it was a hit in L2Cache, return the request back to where it originally came from.
+            // 
+            // Send out the resp to the original SrcUnit -- which is now the DestUnit.
+            sendOutResp_(req->getDestUnit(), req->getInstPtr());
+        }
+        else { // if (req->getCacheState() == L2CacheState::MISS)            
+                
+            // Set Destination for this request to BIU
+            req->setDestUnit(L2ArchUnit::BIU);
+
+            // Handle the miss instruction by storing it aside while waiting 
+            // for lower level memory to return
+            if (miss_pending_buffer_.size() < miss_pending_buffer_size_) {
+
+                ILOG("Storing the CACHE MISS in miss_pending_buffer_");
+                miss_pending_buffer_.push_back(req);
+            }
+            else {
+                sparta_assert("No space in miss_pending_buffer_! Why did the frontend issue push the request onto l2cache_pipeline_?");
+            }
+
+            // Function to check if the request to the given cacheline is present in the miss_pending_buffer_ 
+            auto getCacheLine = [this] (auto reqPtr) { return reqPtr->getInstPtr()->getRAdr() >> shiftBy_; };
+            auto is_addr_present = [req, getCacheLine] (auto reqPtr)
+                            { return (req != reqPtr && getCacheLine(reqPtr) == getCacheLine(req));};
+
+            // Send out the request to BIU for a cache MISS if it is not sent out already
+            auto reqIter = std::find_if(miss_pending_buffer_.begin(), miss_pending_buffer_.end(), is_addr_present);
```

Comment:
Curious, if it does hit, does it make sense to append this request to that miss so you can track how many requests are to that existing miss?  Kinda handy to know...

Reply:
You mean with a separate member of the request that counts the number of misses to the same/originally missed cacheline?

Reply:
Something like that.  You can "append" this request to the found request for reference or for debug.

Reply:
Implemented this. Each request in this buffer points to the next miss to the same cacheline through a member variable `next_req_`. By default/if there is no miss to the same cacheline after the said req in the buffer, `next_req_ = nullptr`.